### DISCUSSION TITLE
How to set an Edit Target to a referenced layer.

### DISCUSSION THREAD

**Discussion by jcoelho88** [Autodesk/maya-usd#3191] [Show and tell]
How does one set an edit target to a reference from a prim?

For the longest time I though you could only set an edit target to a layer that's part of the stages local layer stack, meaning something that might look like this:

![image](https://github.com/Autodesk/maya-usd/assets/88895104/6b01d646-bb89-4328-9047-1b8a905e5375)

In this image I can use the layer editor to pick any of the sublayers that compose the stage to have my edits target.

This is all fine and makes sense but what if your stage was composed using references?

If I structure an asset using references that defines things like the materials and their bindings separate from the mesh and UV's for example then wouldn't it be nice to target those layers to work on specific workflows.

I've only recently learned that this was possible via a thread on the usd-interest group.

You can view the thread here, assuming you have joined the group: https://groups.google.com/g/usd-interest/c/dE-sGU6c8qU/m/9GnqRdoHAAAJ

This next bit is what pointed me to the answer for allowing of edit targeting references.

"_The slick "working in context" feature that USD provides, which also takes care of time-remapping concerns when you are trying to edit animation, is to set up a UsdEditTarget in your Level stage that targets DiningSet.usd through the reference (or whatever arcs bring it in).  Using an EditTarget is very transparent and effortless, and will give you exactly the workflow it sounds like you're looking for.  The problem with EditTargets is that support for setting them up intuitively is lacking in most DCC's, and those that do limit the range of targets to root-layer subLayers (I think).  We developed UsdPrimCompositionQuery to make it easier to "find" the right "target node" and layer with which to configure a UsdEditTarget: the UsdPrimCompositionArc structures that a Query produces give you just what you need, but it takes some learning and either good GUI or knowledge of your asset structure to use it effectively.  You can have a look at [the unit test](https://github.com/PixarAnimationStudios/USD/tree/release/pxr/usd/usd/testenv/testUsdPrimCompositionQuery.testenv) for some pointers, hopefully._ "

So for anyone looking for an example on how this might work here is something I've tested.

```python
# Targeting reference layers
from pxr import Usd, Sdf

# Open the stage
asset_filepath = "/PATH/TO/ASSET"
asset_stage = Usd.Stage.Open(asset_filepath)

# Or if you already have a mayaUsdProxyShape open in Maya
#import mayaUsd
#maya_stage_node = "|MAYA_STAGE_NODE"
#asset_stage = mayaUsd.ufe.getStage(maya_stage_node)

# I'm assuming you have a way to know the path to the layer you want to set as an edit target
layer_filepath = "/PATH/TO/LAYER"

# Im assuming the references are on the default prim. This will return a list of composition arcs.
default_prim = asset_stage.GetDefaultPrim()
composition_arcs = Usd.PrimCompositionQuery(default_prim).GetCompositionArcs()

# Loop through the composition arc and find the target layer that matches your "layer_filepath"
# Do this and store it and its corresponding target node.
target_layer = None
target_layer_node = None

for arc in composition_arcs:
    # Confirm no unexpected issue like the wrong slashes used in the path or if one is using a search/relative path
    if arc.GetTargetLayer().identifier == layer_filepath:
        target_layer = arc.GetTargetLayer()
        target_layer_node = arc.GetTargetNode()
        
        break

# Create your edit target and set it on the stage
edit_target = Usd.EditTarget(target_layer, target_layer_node )
asset_stage.SetEditTarget(edit_target)

```
When you open the Layer editor after setting the edit target to the refenced layer you will notice that there is no displayed layer that is selected. This is because the Layer Editor can't display things that are references. 

![image](https://github.com/Autodesk/maya-usd/assets/88895104/7c4700ad-992f-4d8e-8894-11cc5d51afa6)

Any changes you make now should be targeting what you set as the edit target. To confirm this, write out the layer to the script editor, then make a change and write it out again, you should see your new opinions the second time around.

Here is how to write the layer to the script editor and a few other things to test.

```python
# Write the layer out to the script editor
target_layer.ExportToString()

# If you want to mute your layer
target_layer.SetMuted(True)

# Unmute
target_layer.SetMuted(False)

# Save
target_layer.Save()

# Export to a new layer
target_layer.Export("/NEW/PATH/FOR/LAYER")

# If you want to reload the layer and get rid of your changes before you save or export.
target_layer.Reload()
```

This may not be something everyone will take advantage of but I hope it can maybe unlock some workflows that may have not been considered.

Cheers,

Jason Coelho

**Comment by santosd** [comment]
This is great stuff @jcoelho88 thanks for sharing!

**Comment by mathieubertrand-fs** [comment]
Is there a plan to add the functionality to set the edit target across reference arcs in the USD Layer Editor or another bit of UI? @seando-adsk

**Reply by jcoelho88** [reply]
It would be nice if this was the case.

We are probably going to do a custom tool ourselves to make it easier to switch the edit target to other composition arcs. There is one feature that comes with the USD Layer Editor that would be nice to get.

Saving the Maya file with layers that are dirty and keeping the edits in the scene to later apply on the next open of the maya file.

![image](https://github.com/Autodesk/maya-usd/assets/88895104/a88371ad-87d6-4b48-9995-a174634c9e58)

As of right now the USD layer manager can't save the edits done on a layer not recognized in the Layer editor to the Maya file. Meaning you'd have to save or export the layer in all cases, this is not always something we would want to do and would prefer to be able to save the edits to the Maya scene.

I've tried what I can to get around this but have not had any luck and from what I gather the Layer Manager is the only way to save the edits as text in the Maya file:

https://github.com/Autodesk/maya-usd/blob/dev/lib/mayaUsd/nodes/Layer_Saving_Docs.md

Recognizing across arcs and saving the edits to the Maya file would be awesome.

Cheers,

Jason Coelho

**Comment by wallworm** [comment]
We currently do not have plans to support this, but your discussion has brought it up for discussion on the team. Do you have thoughts on the UX for this workflow? With the current layer editor, the UX is manageable with a top-level layering. I can imagine that very complex scenes with large numbers of references _might_ be challenging to accommodate in the layer editor. Do you have some proposed workflow designs?

**Reply by jcoelho88** [reply]
Hi @wallworm,

On my end this is still early, and I agree that the layer editor will get overly complex as the number of references will likely be very large.

Since the idea is that we want to be targeting a reference (or other composition arc) on a prim then you'd probably only ever need to explicitly show the composition arcs of a prim(s) you have selected.

I'm no expert on UX and I will bring this to our team member who is but, here's one idea.

Have a right-click option like you do for Variant Sets, Deactivate Prim etc and include an "Open Composition Arc". 

A window very similar to the Layer editor opens with the layers that are eligible to be be targeted for that prim arc and the user can switch between the layers just like the current Layer editor.

Like the Layer Editor, unsaved changes to the edit targets should be able to be saved to the Maya scene.

I've no doubt there are better ways and when we get the chance to go over it on our side I can update this post.

Also as an aside, talking about the Layer Editor, it would be nice to have an Export option to write the layer to a new file.

Cheers,

Jason Coelho

**Reply by mathieubertrand-fs** [reply]
I agree this UI will need to be available per prim as your navigate the hierarchy. Unsaved changes should definitely be saved with the maya scene yes.

**Reply by wallworm** [reply]
@jcoelho88 It's been a year since this thread has been updated. I would like to get our team to start solving some of these long-outstanding workflow gaps around composition arcs outside of the local layer stack (including variant, inherits, referenced layers, etc). Above you mentioned that you might go over the topic with your team and update the post. Do you (and anyone else in this thread) have any new insights on UX desires and needs you can share with us? I'd like to revive this topic with those customers having interest in this area.

**Reply by jcoelho88** [reply]
Hi @wallworm,

There is still interest on my end but I do not have any new UX insights to share.

Main points are still the same.

- Being able to target other composition arcs, per prim in most cases. 
- Saving edits to the Maya File instead of saving to the layer itself if the person does not want to push the changes to the layer yet. 
- The option to save the edited layer to a new one instead of overriding it.

An example of this (depends on how you structure your layers), say I have a reference in a shot to an asset that itself is made of other references, (geometry, materials etc.) and I notice that a change to a shader should happen to the asset, I would have the power to target the material layer for the asset via the prim selection and then make an edit to the shader and decide if I either write over that layer or save as something else, possible a new version if you are setup for them.

I'm not saying the above example is correct or good but an option that composition targeting allows.

Cheers,

Jason Coelho

**Reply by wallworm** [reply]
Thank you for this input.
